Gems/Atom/RHI/Code/Source/RHI/MultiDeviceQueryPool.cpp

```diff
@@ -0,0 +1,249 @@
+        }
+
+        return resultCode;
+    }
+
+    ResultCode MultiDeviceQueryPool::InitQuery(MultiDeviceQuery* query)
+    {
+        return InitQuery(&query, 1);
+    }
+
+    ResultCode MultiDeviceQueryPool::InitQuery(MultiDeviceQuery** queries, uint32_t queryCount)
+    {
+        AZ_Assert(queries, "Null queries");
+        auto resultCode{ ResultCode::Success };
+
+        for (auto& [deviceIndex, deviceQueryPool] : m_deviceQueryPools)
+        {
+            AZStd::vector<RHI::Ptr<Query>> deviceQueries(queryCount);
+            AZStd::vector<Query*> rawDeviceQueries(queryCount);
+            for (auto index{ 0u }; index < queryCount; ++index)
+            {
+                deviceQueries[index] = RHI::Factory::Get().CreateQuery();
+                rawDeviceQueries[index] = deviceQueries[index].get();
+            }
+
+            resultCode = deviceQueryPool->InitQuery(rawDeviceQueries.data(), queryCount);
+
+            for (auto index{ 0u }; index < queryCount; ++index)
+            {
+                queries[index]->m_deviceQueries[deviceIndex] = deviceQueries[index];
+            }
+
+            if (resultCode != ResultCode::Success)
+            {
+                break;
+            }
+        }
+
+        if (resultCode == ResultCode::Success)
+        {
+            for (auto index{ 0u }; index < queryCount; ++index)
+            {
+                MultiDeviceResourcePool::InitResource(
+                    queries[index],
+                    []()
+                    {
+                        return ResultCode::Success;
+                    });
+            }
+        }
+
+        return resultCode;
+    }
+
+    ResultCode MultiDeviceQueryPool::ValidateQueries(MultiDeviceQuery** queries, uint32_t queryCount)
+    {
+        if (!queryCount)
+        {
+            AZ_Error("RHI", false, "MultiDeviceQuery count is 0");
+            return RHI::ResultCode::InvalidArgument;
+        }
+
+        for (uint32_t i = 0; i < queryCount; ++i)
+        {
+            auto* query = queries[i];
+            if (!query)
+            {
+                AZ_Error("RHI", false, "Null query %i", i);
+                return RHI::ResultCode::InvalidArgument;
+            }
+
+            if (query->GetQueryPool() != this)
+            {
+                AZ_Error("RHI", false, "MultiDeviceQuery does not belong to this pool");
+                return RHI::ResultCode::InvalidArgument;
+            }
+        }
+        return ResultCode::Success;
+    }
+
+    ResultCode MultiDeviceQueryPool::GetResults(MultiDeviceQuery* query, uint64_t* result, uint32_t resultsCount, QueryResultFlagBits flags)
+    {
+        auto resultCode{ ResultCode::Success };
+
+        for (auto& [deviceIndex, deviceQueryPool] : m_deviceQueryPools)
+        {
+            auto deviceQuery{ query->m_deviceQueries[deviceIndex].get() };
+            auto deviceResult{ result + (deviceIndex * resultsCount) };
```

Comment:
The current version foresees that a user has to allocate `resultsCount * deviceCount` in the result array and we write the corresponding results from a device then to the corresponding memory position (i.e. `startOffset + (deviceIndex * resultsCount)`). 
By compacting it immediately would make it more difficult for the user to query then results for a specific device then later (i.e. a user would need to iterate through all devices to find the one with the corresponding index instead of directly indexing).
I would consider that less clear and more difficult to work with.

Reply:
Ah, now I see your point, that is not optimal, yes (although I am not sure how many devices would typically reside in a work station or server which would be addressable from a single engine instance, so the overhead would not be that much I would guess). 
But thinking about it, it probably would actually be best to change the API here to request results from a specific device, that way a user would then automatically have the correct mapping between device and its results and could also query results for all devices by simply looping over them.
That would just move the loop to the user but also would give the user the mapping directly.
Another option would be to change the API to return a map, mapping from device index to results for this device.

Reply:
We discussed to have the API not be device specific as much as possible for the moment and adapt it at a later point if necessary or beneficial.

However, this is a special case where it is a bit unclear on how to best achieve this. We also discussed passing in a map of `result*` that maps from the device index to preallocated result arrays, but decided for this version since it most closely resembles the existing API of the single device class. Let us know what you think and if (and how) we should change the API.